Clang-tidy invalid suggestion: llvm-twine-local

With clang-tidy built at current head and applying it on MLIR:

```
$ clang-tidy -p build mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp --checks=-*,llvm-twine-local -fix -fix-errors
```

```
diff --git a/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp b/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp
index 453b9bc1b58b..78402e5ea896 100644
--- a/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp
+++ b/mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp
@@ -416,7 +416,7 @@ llvm::LaunchKernel::createKernelLaunch(mlir::gpu::LaunchFuncOp op,
 
   // Load the kernel function.
   StringRef moduleName = op.getKernelModuleName().getValue();
-  Twine moduleIdentifier = getModuleIdentifier(moduleName);
+  std::string moduleIdentifier = (getModuleIdentifier(moduleName)).str();
   Value *modulePtr = module.getGlobalVariable(moduleIdentifier.str(), true);
   if (!modulePtr)
     return op.emitError() << "Couldn't find the binary: " << moduleIdentifier;
```


```
mlir/lib/Target/LLVMIR/Dialect/GPU/SelectObjectAttr.cpp:420:64: error: no member named 'str' in 'std::basic_string<char>'
  420 |   Value *modulePtr = module.getGlobalVariable(moduleIdentifier.str(), true);
      |                                               ~~~~~~~~~~~~~~~~ ^
1 error generated.
```

--- a/clang-tools-extra/docs/clang-tidy/checks/llvm/twine-local.rst
+++ b/clang-tools-extra/docs/clang-tidy/checks/llvm/twine-local.rst
@@ -14,3 +14,21 @@ should be generally avoided.
   // becomes
 
   static std::string Moo = (Twine("bark") + "bah").str();
+
+The ``llvm::Twine`` does not own the memory of its contents, so it is not
+recommended to use ``Twine`` created from temporary strings or string literals.
+
+.. code-block:: c++
+
+  static Twine getModuleIdentifier(StringRef moduleName) {
+    return moduleName + "_module";
+  }
+  void foo() {
+    Twine result = getModuleIdentifier(std::string{"abc"} + "def");
+    // temporary std::string is destroyed here, result is dangling
+  }
+
+After applying this fix-it hints, the code will use ``std::string`` instead of
+``Twine`` for local variables. However, ``llvm::Twine`` has lots of methods that
+are incompatible with ``std::string``, so the user may need to adjust the code
+manually after applying the fix-it hints.

--- a/clang-tools-extra/docs/clang-tidy/checks/llvm/twine-local.rst
+++ b/clang-tools-extra/docs/clang-tidy/checks/llvm/twine-local.rst
@@ -15,7 +15,7 @@ should be generally avoided.
 
   static std::string Moo = (Twine("bark") + "bah").str();
 
-The ``llvm::Twine`` does not own the memory of its contents, so it is not
+The ``Twine`` does not own the memory of its contents, so it is not
 recommended to use ``Twine`` created from temporary strings or string literals.
 
 .. code-block:: c++
@@ -29,6 +29,6 @@ recommended to use ``Twine`` created from temporary strings or string literals.
   }
 
 After applying this fix-it hints, the code will use ``std::string`` instead of
-``Twine`` for local variables. However, ``llvm::Twine`` has lots of methods that
+``Twine`` for local variables. However, ``Twine`` has lots of methods that
 are incompatible with ``std::string``, so the user may need to adjust the code
 manually after applying the fix-it hints.

